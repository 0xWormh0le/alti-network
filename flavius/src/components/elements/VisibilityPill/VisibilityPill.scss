@import 'styles/_shared.scss';

.VisibilityPillLoading {
  width: 70px;
  height: 20px;
}
