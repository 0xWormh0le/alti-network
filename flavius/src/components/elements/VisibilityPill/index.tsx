export * from './VisibilityPill'

export { default } from './VisibilityPill'
