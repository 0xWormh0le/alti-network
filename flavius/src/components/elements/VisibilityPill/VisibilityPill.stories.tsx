import React from 'react'
import { Story, Meta } from '@storybook/react/types-6-0'
import VisibilityPill, { VisibilityPillProps } from './VisibilityPill'

export default {
  title: 'Elements/VisibilityPill',
  component: VisibilityPill
} as Meta

const Template: Story<VisibilityPillProps> = (args) => <VisibilityPill {...args} />

export const External = Template.bind({})
External.args = {
  visibility: 'external'
}

export const Internal = Template.bind({})
Internal.args = {
  visibility: 'internal'
}

export const Sensitive = Template.bind({})
Sensitive.args = {
  visibility: 'sensitive'
}
