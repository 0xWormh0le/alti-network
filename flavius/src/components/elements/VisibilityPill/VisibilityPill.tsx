import React from 'react'
import ContentLoader from 'react-content-loader'
import Tooltip from 'components/widgets/Tooltip'
import Marker, { MarkerType } from 'components/elements/Marker'
import withContentLoader from 'hocs/withContentLoader'
import UI_STRINGS from 'util/ui-strings'
import './VisibilityPill.scss'

export const sharingIndicatorLabelMapping = (visibility: LinkVisibility) => {
  switch (visibility) {
    case 'none':
    case 'group':
    case 'user':
      return UI_STRINGS.VISIBILITY.NONE
    case 'internal':
      return UI_STRINGS.VISIBILITY.INTERNAL
    case 'internal_discoverable':
      return UI_STRINGS.VISIBILITY.INTERNAL_DISCOVERABLE
    case 'external':
      return UI_STRINGS.VISIBILITY.EXTERNAL_LINK_SHARING
    case 'external_discoverable':
      return UI_STRINGS.VISIBILITY.EXTERNAL_DISCOVERABLE
    case 'unknown':
      return UI_STRINGS.VISIBILITY.UNKNOWN
    case 'sensitive':
      return UI_STRINGS.VISIBILITY.SENSITIVE_CONTENT
    default:
      throw new Error(`Link visibility can never be ${visibility}`)
  }
}

export const sharingIndicatorTooltipMapping = (visibility: LinkVisibility) => {
  switch (visibility) {
    case 'none':
    case 'group':
    case 'user':
    case 'unknown':
      return ''
    case 'internal':
    case 'internal_discoverable':
      return UI_STRINGS.VISIBILITY.INTERNAL_TOOLTIP
    case 'external':
    case 'external_discoverable':
      return UI_STRINGS.VISIBILITY.EXTERNAL_TOOLTIP
    case 'sensitive':
      return UI_STRINGS.VISIBILITY.SENSITIVE_CONTENT
    default:
      throw new Error(`Link visibility can never be ${visibility}`)
  }
}

export interface VisibilityPillProps {
  visibility: LinkVisibility
}

const VisibilityPill: React.FC<VisibilityPillProps> = ({ visibility }) => {
  const pillLabel = sharingIndicatorLabelMapping(visibility)
  const tooltipText = sharingIndicatorTooltipMapping(visibility)

  if (tooltipText !== '') {
    return (
      <Tooltip text={tooltipText} id={`tooltip-visibility-${pillLabel}`}>
        <Marker type={visibility as MarkerType} />
      </Tooltip>
    )
  } else {
    return <Marker type={visibility as MarkerType} />
  }
}

const VisibilityPillLoading = () => (
  <ContentLoader
    backgroundColor='#f0f0f0'
    foregroundColor='#f7f7f7'
    height={20}
    width={70}
    className='VisibilityPillLoading'
    uniqueKey='VisibilityPillLoading'>
    <rect x={0} y={0} width={70} height={20} />
  </ContentLoader>
)

export default withContentLoader(VisibilityPillLoading)(VisibilityPill)
