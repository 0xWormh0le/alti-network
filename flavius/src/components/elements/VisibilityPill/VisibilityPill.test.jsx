import React from 'react'
import VisibilityPill from './VisibilityPill'
import { render } from '@testing-library/react'

it('renders correctly', () => {
  const { container } = render(<VisibilityPill visibility='internal' />)
  expect(container).toMatchSnapshot()
})

it('renders name and classname for `internal` status', () => {
  const { getByText } = render(<VisibilityPill visibility='internal' />)

  const statusElement = getByText(/internal/i)

  expect(statusElement).toBeInTheDocument()
  expect(statusElement).toHaveClass('Marker__internal')
})

it('renders name and classname for `external` status', () => {
  const { getByText } = render(<VisibilityPill visibility='external' />)

  const statusElement = getByText(/external/i)

  expect(statusElement).toBeInTheDocument()
  expect(statusElement).toHaveClass('Marker__external')
})

it('renders name and classname for `internal_discoverable` status', () => {
  const { getByText } = render(<VisibilityPill visibility='internal_discoverable' />)

  const statusElement = getByText(/internal & discoverable/i)

  expect(statusElement).toBeInTheDocument()
  expect(statusElement).toHaveClass('Marker__internal_discoverable')
})

it('renders name and classname for `external_discoverable` status', () => {
  const { getByText } = render(<VisibilityPill visibility='external_discoverable' />)

  const statusElement = getByText(/external & discoverable/i)

  expect(statusElement).toBeInTheDocument()
  expect(statusElement).toHaveClass('Marker__external_discoverable')
})

describe('renders None visibility', () => {
  const statuses = ['none', 'user', 'group']

  statuses.forEach((status) => {
    it(`renders name and classname for ${status} status`, () => {
      const { getByText } = render(<VisibilityPill visibility={status} />)

      const statusElement = getByText(/No Link Sharing/i)

      expect(statusElement).toBeInTheDocument()
      expect(statusElement).toHaveClass(`Marker__${status}`)
    })
  })
})

it('renders tooltip correctly', () => {
  const { container } = render(<VisibilityPill visibility='external_discoverable' />)
  expect(container.querySelector('.Marker')).toHaveClass('trigger')
})
