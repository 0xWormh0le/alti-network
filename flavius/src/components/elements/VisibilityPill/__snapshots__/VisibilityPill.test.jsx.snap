// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <span
    class="Typography Typography--body trigger Marker Marker__internal"
  >
    Internal Link Sharing
  </span>
</div>
`;
